draft-ietf-tls-tls13.md

```diff
@@ -3390,14 +3424,29 @@ TLS-compliant application MUST implement the following TLS extensions:
 
   * Signature Algorithms ("signature_algorithms"; {{signature-algorithms}})
   * Negotiated Groups ("supported_groups"; {{negotiated-groups}})
+  * Client Key Share ("client_key_share"; {{client-key-share}})
+  * Pre-Shared Key Extension ("pre_shared_key"; {{pre-shared-key-extension}})
+  * Server Name Indication ("server_name"; Section 3 of {{RFC6066}})
+
+All implementations MUST send and use these extensions when offering
+applicable cipher suites:
+
+  * "signature_algorithms" is REQUIRED for certificate authenticated cipher suites
+  * "supported_groups" and "client_key_share" are REQUIRED for DHE or ECDHE cipher suites
+  * "pre_shared_key" is REQUIRED for PSK cipher suites
+
+When negotiating use of applicable cipher suites, endpoints MUST abort the
+connection with a "missing_extension" alert if the required extension was
+not provided. Any endpoint that receives any invalid combination of cipher
+suites and extensions MAY abort the connection with a "missing_extension"
+alert, regardless of negotiated parameters.
 
-All implementations MUST use the "signature_algorithms" extension when
-offering and negotiating certificate authenticated cipher suites.
-All implementations MUST use the "supported_groups" extension when
-offering and negotiating DHE or ECDHE cipher suites.
+Additionally, all implementations MUST support use of the "server_name"
+extension with applications capable of using it.
+Servers MAY require clients to send a valid "server_name" extension.
+Servers requiring this extension SHOULD respond to a ClientHello
+lacking a "server_name" extension with a fatal "missing_extension" alert.
```

Comment:
SNI is already a MAY require and SHOULD send:
https://tlswg.github.io/tls13-spec/#server-certificate

There's just no statement of what to do when it's required and not given.

Encrypted SNI is essentially in the current draft now. The EncryptedExtensions message is now considered mandatory to send for clients for 0-RTT, so just sticking SNI in there should do the trick. I think the consensus was that rearranging the world to attempt to get it encrypted in a normal first-flight is a mess and not worth the effort. If we can actually get the ServerConfiguration to be easily deployable through other channels, this could allow for first-connect encrypted SNI via 0-RTT, which would be great. The text here works with all this just fine as it doesn't require a new version of the extension. One consideration might be to add a flag to the server config as to whether or not SNI should be encrypted, just in case someone would rather have 1-RTT fallback for SNI instead of expecting 0-RTT to work as announced.